library/L1_Peripheral/stm32f10x/pwm.hpp

```diff
@@ -0,0 +1,446 @@
+          .Set(kPreloadEnableHigh)
+          .Save();
+        break;
+      case 3:
+        bit::Register(&pwm_pin_.registers->CCMR2)
+          .Set(kPreloadEnableLow)
+          .Save();
+        break;
+      case 4:
+        bit::Register(&pwm_pin_.registers->CCMR2)
+          .Set(kPreloadEnableHigh)
+          .Save();
+        break;
+      }
+
+      // Output channel enable
+      switch (pwm_pin_.channel)
+      {
+      case 1:
+        bit::Register(&pwm_pin_.registers->CCER)
+          .Set(kChannel1OutputEnable)
+          .Save();
+        break;
+      case 2:
+        bit::Register(&pwm_pin_.registers->CCER)
+          .Set(kChannel2OutputEnable)
+          .Save();
+        break;
+      case 3:
+        bit::Register(&pwm_pin_.registers->CCER)
+          .Set(kChannel3OutputEnable)
+          .Save();
+        break;
+      case 4:
+        bit::Register(&pwm_pin_.registers->CCER)
+          .Set(kChannel4OutputEnable)
+          .Save();
+        break;
+      }
+
+      // Main output enable
+      if (pwm_pin_.id == SystemController::Peripherals::kTimer1)
+      {
+        bit::Register(&pwm_pin_.registers->BDTR)
+          .Set(kMainOutputEnable)
+          .Save();
+      }
+
+      // Timer enable
+      bit::Register(&pwm_pin_.registers->CR1)
+        .Set(kTimerEnable)
+        .Save();
+    }
+    else
+    {
+      // Preload enable
+      switch (pwm_pin_.channel)
+      {
+      case 1:
+        bit::Register(&pwm_pin_.registers->CCMR1)
+          .Clear(kPreloadEnableLow)
+          .Save();
+        break;
+      case 2:
+        bit::Register(&pwm_pin_.registers->CCMR1)
+          .Clear(kPreloadEnableHigh)
+          .Save();
+        break;
+      case 3:
+        bit::Register(&pwm_pin_.registers->CCMR2)
+          .Clear(kPreloadEnableLow)
+          .Save();
+        break;
+      case 4:
+        bit::Register(&pwm_pin_.registers->CCMR2)
+          .Clear(kPreloadEnableHigh)
+          .Save();
+        break;
+      }
+
+      // Output channel enable
+      switch (pwm_pin_.channel)
+      {
+      case 1:
+        bit::Register(&pwm_pin_.registers->CCER)
+          .Clear(kChannel1OutputEnable)
+          .Save();
+        break;
+      case 2:
+        bit::Register(&pwm_pin_.registers->CCER)
+          .Clear(kChannel2OutputEnable)
+          .Save();
+        break;
+      case 3:
+        bit::Register(&pwm_pin_.registers->CCER)
+          .Clear(kChannel3OutputEnable)
+          .Save();
+        break;
+      case 4:
+        bit::Register(&pwm_pin_.registers->CCER)
+          .Clear(kChannel4OutputEnable)
+          .Save();
+        break;
+      }
+
+      // Main output enable
+      if (pwm_pin_.id == SystemController::Peripherals::kTimer1)
+      {
+        bit::Register(&pwm_pin_.registers->BDTR)
+          .Clear(kMainOutputEnable)
+          .Save();
+      }
+
+      // Timer enable
+      bit::Register(&pwm_pin_.registers->CR1)
+        .Clear(kTimerEnable)
+        .Save();
+    }
```

Comment:
Wait does this make sense though? If I want to disable just 1 channel, why do I have to turn off the whole timer and also lose all of my other channels. This doesn't seem like the behavior you'd expect from disabling a single channel of PWM.

Reply:
Yea, that's a good point. I'll fix that.